dag_in_context/src/lib.rs

```diff
@@ -95,12 +104,72 @@ pub(crate) fn print_with_intermediate_vars(termdag: &TermDag, term: Term) -> Str
     printed
 }
 
+// Takes in a termdag and a shared cache referencing the termdag, which can be re-used for multiple exprs
+fn global_cached_print_expr_with_intermediate_helper(
```

Comment:
This has a pretty confusing name- I think you should inline this function so it doesn't exist